[WIP] feature: Added Confiant AdQuality module

IvanTsymbal opened the pull request:
Current PR introduces new public Module which can communicate to the specific Confiant's server and filter Auction Bids responses if they contain some Security issues after scan process (https://www.confiant.com/).

Initial aim of the PR is to receive a feedback if we allowed to extend AuctionInterface of the AllProcessedBidResponses and general feedback about the module implementation itself.

bretg commented:
@IvanTsymbal - please summarize what this statement means:

> extend AuctionInterface of the AllProcessedBidResponses

Is this an impact at the requirements level?

IvanTsymbal commented:
> @IvanTsymbal - please summarize what this statement means:
> 
> > extend AuctionInterface of the AllProcessedBidResponses
> 
> Is this an impact at the requirements level?

@bretg In order for Confiant to do a proper scan, we need to pass the bid request to our server on the Prebid’s stage with name All Processed Bids Responses https://docs.prebid.org/prebid-server/developers/add-a-module.html#2-understand-the-endpoints-and-stages.
We will achieve that by enriching the auction context model with the reference to the bidRequest object - so we are adding a new field to the interface https://github.com/prebid/prebid-server-java/pull/2326/files#diff-ba1fb5b53e3206ca91e58fc03d6f56b86ce77cd5d6515ddb81152fd566ef219dR15

bretg commented:
@IvanTsymbal - I've opened an internal ticket at prioritized review of this for the team.

There's a delay in many projects right now because GPP is time-sensitive.

And1sS commented on extra/modules/confiant-ad-quality/src/main/java/org/prebid/server/hooks/modules/com/confiant/adquality/core/RedisScanStateChecker.java:
```diff
@@ -0,0 +1,36 @@
+
+public class RedisScanStateChecker {
+
+    private boolean isScanDisabled = true;
```
This has to be at least marked as `volatile`, as this will be accessed from multiple threads. JMM does not have any guarantees with current approach. Unsynchronized reads/writes can be just unseen(even though on modern architectures this should not be the case).

And1sS commented on extra/modules/confiant-ad-quality/src/main/java/org/prebid/server/hooks/modules/com/confiant/adquality/core/RedisClient.java:
```diff
@@ -0,0 +1,84 @@
+    public BidsScanResult submitBids(RedisBidsData bids) {
+        if (jedis != null && bids.getBresps().size() > 0) {
+            try {
+                final String submitHash = jedis.get("function_submit_bids");
```
Why don't you provide function code? Also, what is the point in using redis here? This has to be a backend application.

And1sS commented on src/main/proto/openrtb-xapi.proto:
```diff
@@ -0,0 +1,565 @@
+//
```
This should not be a part of repository, as proto files are downloaded at build time.

And1sS commented on src/main/proto/openrtb.proto:
```diff
@@ -0,0 +1,2018 @@
+syntax = "proto2";
```
This one too.

And1sS commented on extra/modules/confiant-ad-quality/src/test/java/org/prebid/server/hooks/modules/com/confiant/adquality/core/BidsMapperTest.java:
```diff
@@ -0,0 +1,77 @@
+    @Test
+    public void shouldMapBidResponsesToRedisBids() {
+        // given
+        BidRequest bidRequest = BidRequest.builder()
```
Please, use `final` keyword everywhere. Also, please, check for similar occurences.

IvanTsymbal commented:
@And1sS Thanks a lot that you took a look into current PR.
I hope I've done required updates according to your notes - changed code style for braces, added missing final keywords, volatile access, removed generated proto files.
Also did replacement to `vertx` instead of blocking operations and jedis lib (was not very experienced in that technology so hope now that's fine).
Regarding usage of the redis functions - it's an internal mechanism and part of the Confiant architecture which we use internally.

Besides that I still would like to hear if my changes for core prebid hook are allowed (new property for the context):
https://github.com/prebid/prebid-server-java/pull/2326/files#diff-ba1fb5b53e3206ca91e58fc03d6f56b86ce77cd5d6515ddb81152fd566ef219dR15

Thanks.

CTMBNara commented:
> Besides that I still would like to hear if my changes for core prebid hook are allowed (new property for the context): https://github.com/prebid/prebid-server-java/pull/2326/files#diff-ba1fb5b53e3206ca91e58fc03d6f56b86ce77cd5d6515ddb81152fd566ef219dR15

You can use `auctionInvocationContext.auctionContext().getBidRequest` instead.

CTMBNara commented:
> How did I miss that initially?!.. Thanks a lot for your hint. Have changed.

It was added recently: https://github.com/prebid/prebid-server-java/pull/2308/files#diff-ba1fb5b53e3206ca91e58fc03d6f56b86ce77cd5d6515ddb81152fd566ef219dR9

You should also pay attention to the fact that you want to pass a request with an unmasked user to a third-party service. This may lead to legal problems. Worth checking with @bretg .

IvanTsymbal commented:
> > How did I miss that initially?!.. Thanks a lot for your hint. Have changed.
> 
> It was added recently: https://github.com/prebid/prebid-server-java/pull/2308/files#diff-ba1fb5b53e3206ca91e58fc03d6f56b86ce77cd5d6515ddb81152fd566ef219dR9
> 
> You should also pay attention to the fact that you want to pass a request with an unmasked user to a third-party service. This may lead to legal problems. Worth checking with @bretg .

Thanks @CTMBNara .
@bretg will it be a problem?..
Potentially Confiant needs only 2 properties to work with: Geo info object and Consent string. Will it help if I create specific mapper which does copy of the User object only with current properties?. It may have small influence on the performance because requires additional process of copying, but as an alternative may help.

bretg commented:
@IvanTsymbal - modules need to align with the new "Activity Infrastructure". See https://docs.prebid.org/prebid-server/features/pbs-activitycontrols.html for background.

Specifically in your case, if you're passing geo information to your backend, your module needs to do the following:

1. Call the PBS-core function asking for permission for the `transmitGeo` activity. @SerhiiNahornyi or @CTMBNara  should be able to help with the syntax here. Not clear we've gotten around to the docs yet.
2. If you get permission, cool - send away.
3. If you don't have permission, you need to round off the user's IP addresses (v4 and v6) and lat/long before sending. 
    1. IPv4: replace the last octet with .0
    2. IPv6: See https://docs.prebid.org/prebid-server/features/pbs-privacy.html#anonymizing-ipv6-addresses
    3. lat/long: reduce the number of decimal places to 2

Notes:
- if you send user IDs or first party data to your endpoint we need to talk about other activity controls
- If what you need is pre-rounding geolookup results for country and region, I would suggest the right solution is to trust what PBS gives you in device.geo. Turns out that's still post-rounding, but will change someday.

CTMBNara commented:
@IvanTsymbal Hi. Below you can find an example using the `ActivityInfrastructure`.

```
        final AuctionContext auctionContext = invocationContext.auctionContext();
        final BidRequest bidRequest = auctionContext.getBidRequest();

        // ActivityCallPayload creation will soon be improved to make it more "handy",
        // but that will most likely be after we merge this pull request :)
        final ActivityCallPayload activityCallPayload = BidRequestActivityCallPayload.of(
                ActivityCallPayloadImpl.of(ComponentType.GENERAL_MODULE, "MODULE_NAME"),
                bidRequest);
        final boolean disallowTransmitGeo = !auctionContext.getActivityInfrastructure()
                .isAllowed(Activity.TRANSMIT_GEO, activityCallPayload);

        // PrivacyEnforcementService need to be injected
        // mask user
        final User maskedUser = privacyEnforcementService
                .maskUserConsideringActivityRestrictions(bidRequest.getUser(), true, disallowTransmitGeo);
        // mask device if device.geo is required
        final Device maskedDevice = privacyEnforcementService
                .maskDeviceConsideringActivityRestrictions(bidRequest.getDevice(), true, disallowTransmitGeo);

        // masked BidRequest that this module can use
        final BidRequest maskedBidRequest = BidRequest.builder()
                .user(maskedUser)
                // provide device object if device.geo is required
                .device(maskedDevice)
                .build();
```

IvanTsymbal commented:
Hi. Thanks @CTMBNara - I've added masking for Geo information as you told me - https://github.com/prebid/prebid-server-java/pull/2326/files#diff-d7b042faf034319b49bde325754fd9f32e7851d7223a6f822f08fd5bb1733456R81. Tests which has to verify it also are there.

@bretg - according to methods code which does masking - I see that it removes such info, as IDs, from User object.
https://github.com/prebid/prebid-server-java/blob/master/src/main/java/org/prebid/server/auction/PrivacyEnforcementService.java#L743

What else in the area of "activity controls" do I need to check? Thanks.

bretg commented:
> What else in the area of "activity controls" do I need to check? 

If you're not passing the ORTB `user` object, then transmitGeo is the only activity you need to check.

IvanTsymbal commented:
@CTMBNara, @And1sS  do you in a team see other code issues which I have to fix?

If not - @bretg how should we proceed further? Should we plan a date when we can merge a module and do some real testing? Thanks.

bretg commented:
@IvanTsymbal - what do you mean by "do some real testing"? Do you run a Prebid Server yourselves? Do you have a customer in mind?

IvanTsymbal commented:
@bretg currently main focus is for Raptive (aka Cafemedia). The Magnite is running their Prebid servers.
I've already confirmed that Confiant will setup a meet to discuss the test procedure with the managers.

bretg commented:
ok @IvanTsymbal - we can support that test after this is merged and released. I'll need you to draft a module document like https://docs.prebid.org/prebid-server/pbs-modules/ortb2-blocking.html. If you can open a PR against https://github.com/prebid/prebid.github.io, great, otherwise, just write the content and I'll post the doc.

IvanTsymbal commented:
> ok @IvanTsymbal - we can support that test after this is merged and released. I'll need you to draft a module document like https://docs.prebid.org/prebid-server/pbs-modules/ortb2-blocking.html. If you can open a PR against https://github.com/prebid/prebid.github.io, great, otherwise, just write the content and I'll post the doc.

Sure. I will prepare docs and create related PR.

And1sS commented on extra/modules/confiant-ad-quality/README.md:
```diff
@@ -0,0 +1,72 @@
+# Overview
+
+This module allows obtains all bid responses for any given auction, send the results to Confiant server and find possible security issues inside bid responses.
```
More like: "This module obtains all bid responses for any given auction, sends the results to Confiant server and finds possible security issues inside bid responses."

And1sS commented on extra/modules/confiant-ad-quality/src/main/java/org/prebid/server/hooks/modules/com/confiant/adquality/config/ConfiantAdQualityModuleConfiguration.java:
```diff
@@ -0,0 +1,59 @@
+@PropertySource(
+        value = "classpath:/module-config/confiant-ad-quality.yaml",
+        factory = YamlPropertySourceFactory.class)
+
```
Redundant empty line.

And1sS commented on extra/modules/confiant-ad-quality/src/main/java/org/prebid/server/hooks/modules/com/confiant/adquality/config/ConfiantAdQualityModuleConfiguration.java:
```diff
@@ -0,0 +1,59 @@
+        final RedisScanStateChecker redisScanStateChecker = new RedisScanStateChecker(redisClient, checkInterval, vertx);
+
+        // Initialize Redis connection and scan state check timer
+        final Promise<Void> startClient = Promise.promise();
```
I suggest renaming this to smth like `clientPromise`. Also, please, remove unnecessary comment.

And1sS commented on extra/modules/confiant-ad-quality/src/main/java/org/prebid/server/hooks/modules/com/confiant/adquality/core/BidsMapper.java:
```diff
@@ -0,0 +1,38 @@
+                .stream().map(bidResponse -> RedisBidResponseData
+                        .builder()
+                        .dspId(bidResponse.getBidder())
+                        .bidresponse(BidResponse.builder()
+                                .id(bidRequest.getId())
+                                .cur(bidRequest.getCur().get(0))
+                                .seatbid(Collections.singletonList(SeatBid.builder()
+                                        .bid(bidResponse.getSeatBid().getBids().stream().map(BidderBid::getBid).toList())
+                                        .build()))
```
So, all currencies and seatbids except first will be removed after applying this module? I guess that's not what we want. Correct me if I am wrong.

Also, if you really want to rebuild bid responses, please, extract this to separate method, to reduce nesting

IvanTsymbal replied:
It will not remove anything. After the scan module can only filter responses if some issues exist. It won't modify existing data.

Unfortunately I have to apply mapping because hook receives BidderResponse model and our API requires BidResponse. And as I understand - BidderResponse contains only single SeatBid. If I've missed something - would appreciate help to find what.

Regarding Currency ... For now that's not very important and honestly I'm not 100% about proper value. We expect to see single currency. So I've used main currency from BidRequest. After investigation of the BidderResponse I see that every BidderBid has currency value. And I'm not sure if they are different which result we expect. 

I'm going to move it into separate method.

IvanTsymbal replied:
btw have checked internal Prebid mapping for auctions - https://github.com/prebid/prebid-server-java/blob/master/src/main/java/org/prebid/server/auction/BidResponseCreator.java#L1219
it also takes first currency inside.

And1sS commented on extra/modules/confiant-ad-quality/src/main/java/org/prebid/server/hooks/modules/com/confiant/adquality/core/BidsScanResult.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    public boolean hasIssues() {
+        return result.getValue() != null && !result.getValue().isEmpty() && result.getValue().stream()
```
I think this can be simplified with `Optional`

And1sS commented on extra/modules/confiant-ad-quality/src/main/java/org/prebid/server/hooks/modules/com/confiant/adquality/core/BidsScanResult.java:
```diff
@@ -0,0 +1,43 @@
+        return result.getValue().stream()
+                .map(r -> r.getTagKey() + ": " + (r.getIssues() == null ? "no issues" : r.getIssues().toString()))
+                .toList();
+    }
```
Please, add empty line after method definition.

And1sS commented on extra/modules/confiant-ad-quality/src/main/java/org/prebid/server/hooks/modules/com/confiant/adquality/core/RedisVerticle.java:
```diff
@@ -0,0 +1,103 @@
+import io.vertx.redis.client.RedisOptions;
+import org.prebid.server.hooks.modules.com.confiant.adquality.model.RedisRetryConfig;
+
+public class RedisVerticle extends AbstractVerticle {
```
Why do you need a separate verticle just to start redis on event loop? Consider looking into pbs `ContextRunner` class usages. 
Also, this class should be just a part of redis client class.

And1sS commented on extra/modules/confiant-ad-quality/src/main/java/org/prebid/server/hooks/modules/com/confiant/adquality/core/RedisVerticle.java:
```diff
@@ -0,0 +1,103 @@
+    /**
+     * Attempt to reconnect up to MAX_RECONNECT_RETRIES
+     */
+    private void attemptReconnect(int retry, Handler<AsyncResult<RedisConnection>> handler) {
```
Consider looking into pbs `RetryPolicy` class, it can simplify retry logic.

And1sS commented on extra/modules/confiant-ad-quality/src/main/java/org/prebid/server/hooks/modules/com/confiant/adquality/v1/ConfiantAdQualityBidResponsesScanHook.java:
```diff
@@ -0,0 +1,121 @@
+            AuctionInvocationContext auctionInvocationContext) {
+        final BidRequest bidRequest = getBidRequest(auctionInvocationContext);
+        final List<BidderResponse> responses = allProcessedBidResponsesPayload.bidResponses();
+        final boolean isScanDisabled = redisScanStateChecker.isScanDisabled();
+
+        if (isScanDisabled) {
+            return getResult(new BidsScanResult(OperationResult.empty()), auctionInvocationContext);
+        }
```
Why not abstract this inside redis client class? So that caller will not duplicate/forget to call this logic.

And1sS commented on extra/modules/confiant-ad-quality/src/main/java/org/prebid/server/hooks/modules/com/confiant/adquality/v1/ConfiantAdQualityBidResponsesScanHook.java:
```diff
@@ -0,0 +1,121 @@
+                .build();
+    }
+
+    private Future<InvocationResult<AllProcessedBidResponsesPayload>> getResult(
```
More like `makeResult` or `toInvocationResult`, or smth like that. Also, return type should not be a `Future` as this method doesn't perform any long-running async tasks.

And1sS commented on extra/modules/confiant-ad-quality/src/main/java/org/prebid/server/hooks/modules/com/confiant/adquality/v1/ConfiantAdQualityBidResponsesScanHook.java:
```diff
@@ -0,0 +1,121 @@
+            return getResult(new BidsScanResult(OperationResult.empty()), auctionInvocationContext);
+        }
+
+        final Promise<InvocationResult<AllProcessedBidResponsesPayload>> processed = Promise.promise();
+
+        redisClient.submitBids(BidsMapper.bidResponsesToRedisBids(bidRequest, responses))
+                .onComplete(result -> processed
+                        .complete(getResult(result.result(), auctionInvocationContext).result()));
+
+        return processed.future();
```
After removing `Future` from previous comment, we can simplify this block of code:
```java 
 return redisClient.submitBids(BidsMapper.bidResponsesToRedisBids(bidRequest, responses))
                .map(scanResult -> getResult(scanResult, auctionInvocationContext)));
  ```

And1sS commented on extra/modules/confiant-ad-quality/src/main/java/org/prebid/server/hooks/modules/com/confiant/adquality/core/RedisClient.java:
```diff
@@ -0,0 +1,81 @@
+import java.util.Collections;
+import java.util.List;
+
+public class RedisClient {
```
Also, why do you hardcode redis implementation, looks like this is not RedisClient, this is BidsScanner.

And1sS commented on extra/modules/confiant-ad-quality/src/main/java/org/prebid/server/hooks/modules/com/confiant/adquality/model/RedisBidsData.java:
```diff
@@ -0,0 +1,26 @@
+
+    List<RedisBidResponseData> bresps;
+
+    public String toJson() {
+        try {
+            return new ObjectMapper().writeValueAsString(this);
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException(e);
+        }
+    }
```
You are creating new `ObjectMapper` each time this method is called. 
1) This is heavy.
2) Its not data class responsibility to decide how it should be serialized.

Please, extract ObjectMapper creation and usage to the place where this method is called.

SerhiiNahornyi commented:
@IvanTsymbal Please remove [WIP] from the PR name if you feel PR-ready.
Also FYI some refactoring will be done by the PBS team after PR will be merged.

IvanTsymbal commented:
Thanks @And1sS , @SerhiiNahornyi for your support.

If you see later any important changes which are required - please ping me, I will try to help / resolve them. I guess once we have first test on some pre-prod with real data - will have more feed back about module state.

In situation that PR is merged now - I'm going to manage another repo with module documentation.

IvanTsymbal commented:
> ok @IvanTsymbal - we can support that test after this is merged and released. I'll need you to draft a module document like https://docs.prebid.org/prebid-server/pbs-modules/ortb2-blocking.html. If you can open a PR against https://github.com/prebid/prebid.github.io, great, otherwise, just write the content and I'll post the doc.

Hi, @bretg . I've created initial PR there:
https://github.com/prebid/prebid.github.io/pull/4815
Once have more info - will actualize.
